internal/mdns/mdns.go

```diff
@@ -0,0 +1,225 @@
+	hostIDPretty := hostID.Pretty()
+	txt := []string{hostIDPretty}
+	mdnsService, err := mdns.NewMDNSService(hostIDPretty, s.serviceTag, "", "", int(port), ips, txt)
+	if err != nil {
+		return fmt.Errorf("creating mDNS service: %w", err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: mdnsService})
+	if err != nil {
+		return fmt.Errorf("creating mDNS server: %w", err)
+	}
+	s.server = server
+
+	s.started = true
+	s.stop = make(chan struct{})
+	s.done = make(chan struct{})
+	ready := make(chan struct{})
+
+	go s.run(ready)
+	// It takes a few milliseconds to launch a goroutine
+	// so we wait for the run goroutine to be ready.
+	<-ready
+
+	return nil
+}
+
+// Stop stops the mDNS service and server.
+func (s *Service) Stop() (err error) {
+	if !s.started {
+		return nil
+	}
+
+	defer func() {
+		s.started = false
+	}()
+	close(s.stop)
+	<-s.done
+	return s.server.Shutdown()
+}
+
+func (s *Service) run(ready chan<- struct{}) {
+	defer close(s.done)
+
+	const pollPeriod = time.Minute
+	ticker := time.NewTicker(pollPeriod)
+	defer ticker.Stop()
+
+	handleEntriesReady := make(chan struct{})
+	entriesListeningLoopStop := make(chan struct{})
+	entriesListeningLoopDone := make(chan struct{})
+	entriesCh := make(chan *mdns.ServiceEntry, 16)
+	entriesStartListening := make(chan struct{})
+	entriesStopListening := make(chan struct{})
+
+	go s.handleEntries(handleEntriesReady, entriesListeningLoopStop, entriesListeningLoopDone,
+		entriesStartListening, entriesStopListening, entriesCh)
+	<-handleEntriesReady
+
+	const queryTimeout = 5 * time.Second
+	params := &mdns.QueryParam{
+		Domain:  "local",
+		Entries: entriesCh,
+		Service: s.serviceTag,
+		Timeout: queryTimeout,
+	}
+
+	close(ready)
+
+	for {
+		entriesStartListening <- struct{}{}
+		err := mdns.Query(params)
+		if err != nil {
+			s.logger.Warnf("mdns query failed: %s", err)
+		}
+		entriesStopListening <- struct{}{}
+
+		// Drain the entries channel, we no longer care about entries.
+		for len(entriesCh) > 0 {
+			<-entriesCh
+		}
+
+		select {
+		case <-ticker.C:
+		case <-s.stop:
+			close(entriesListeningLoopStop)
+			<-entriesListeningLoopDone
+			close(entriesCh)
+			close(entriesStartListening)
+			close(entriesStopListening)
+			return
+		}
+	}
+}
+
+func (s *Service) handleEntries(ready chan<- struct{}, stop <-chan struct{}, done chan<- struct{},
+	startListening, stopListening <-chan struct{}, entries <-chan *mdns.ServiceEntry) {
+	defer close(done)
+	close(ready)
+
+	for {
+		// Wait for the start signal to start listening for entries
+		select {
+		case <-startListening:
+		case <-stop:
+			return
+		}
+
+		continueListening := true
+		for continueListening {
+			// Listen for entries until we receive a stop listening signal.
+			select {
+			case entry := <-entries:
+				err := s.handleEntry(entry)
+				if err != nil {
+					s.logger.Warnf("handling mDNS entry: %s", err)
+				}
+			case <-stopListening:
+				continueListening = false
+			case <-stop:
+				return
+			}
+		}
```

Comment:
```suggestion
	listening:
		for {
			// Listen for entries until we receive a stop listening signal.
			select {
			case entry := <-entries:
				err := s.handleEntry(entry)
				if err != nil {
					s.logger.Warnf("handling mDNS entry: %s", err)
				}
			case <-stopListening:
				break listening
			case <-stop:
				return
			}
		}
```

Reply:
This increases indentation :cry: And also labels are not really common in Go code, so I tend to avoid for readers/newcomers. Keeping this as is :wink: